shpc/main/registry/remote.py

```diff
@@ -91,95 +76,81 @@ def override_exists(self, tag):
+                type(self).__name__ + "registry must be a remote path, got %s." % source
+            )
+        self.url = source
+        self.parsed_url = urllib.parse.urlparse(source)
+        self._clone = None
+
+        self.tag = tag
 
         # Cache of remote container metadata
         self._cache = {}
 
         # E.g., subdirectory with registry files
-        self.subdir = kwargs.get("subdir")
-        super().__init__(*args, **kwargs)
-        self._url = self.source
-
-    @classmethod
-    def matches(cls, source):
-        return cls.provider_name in source and source.startswith("http")
+        self.subdir = subdir
 
     @property
-    def source_url(self):
+    def library_url(self):
         """
-        Retrieve a parsed / formatted url, ensuring https and without git.
+        Retrieve the URL of this registry's library (in JSON).
         """
-        url = self.source
-        if not url.startswith("http"):
-            url = "https://%s" % url
-        if url.endswith(".git"):
-            url = url[:-4]
-        return url
-
-    @property
-    def web_url(self):
-        """
-        Retrieve the web url, either pages or (eventually) custom.
-        """
-        parts = self.source_url.split("/")[3:]
-        return "https://%s.%s.io/%s/library.json" % (
-            parts[0],
-            self.provider_name,
-            "/".join(parts[1:]),
-        )
+        raise NotImplementedError
 
     def exists(self, name):
         """
         Determine if a module exists in the registry.
         """
         name = name.split(":")[0]
-        if self._cache and name in self._cache:
-            return True
-        dirname = self.source
-        if self.subdir:
-            dirname = os.path.join(dirname, self.subdir)
-        return os.path.exists(os.path.join(dirname, name))
+        self._update_cache()
+        return name in self._cache
 
     def clone(self, tmpdir=None):
         """
         Clone the known source URL to a temporary directory
+        and return an equivalent local registry (Filesystem)
         """
+        if self._clone:
+            return self._clone
         tmpdir = tmpdir or shpc.utils.get_tmpdir()
 
         cmd = ["git", "clone", "--depth", "1"]
         if self.tag:
             cmd += ["-b", self.tag]
-        cmd += [self._url, tmpdir]
-        self.source = tmpdir
+        cmd += [self.url, tmpdir]
+        if self.subdir:
+            tmpdir = os.path.join(tmpdir, self.subdir)
         try:
             sp.run(cmd, check=True)
         except sp.CalledProcessError as e:
-            raise ValueError("Failed to clone repository {}:\n{}", self.source, e)
-        return tmpdir
+            raise ValueError("Failed to clone repository {}:\n{}", self.url, e)
+        assert os.path.exists(tmpdir)
+        self._clone = Filesystem(tmpdir)
+        return self._clone
+
+    def cleanup(self):
+        """
+        Cleanup the registry
+        """
+        if self._clone:
```

Comment:
I'm someone that regularly nukes my tmpdir at random points "just for funsies" so we check for defined and still existing!

```suggestion
        if self._clone and os.path.exists(self._clone):
```